File: backend/app/services/matching_service.py
from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass
from typing import Dict, List, Tuple

from ..core.cache import CacheProvider
from ..core.config import settings
from ..core.geo import haversine_km
from ..core.sharding import SpatialShardRouter
from ..core.shard_store import ShardExpertStore
from ..db.repository import Repository


KEYWORD_MAP: Dict[str, List[str]] = {
    "Plumbing": ["plumb", "pipe", "leak", "drain", "faucet", "tap", "sink", "pipeline", "bathroom"],
    "Electrical": ["electrical", "wire", "wiring", "switch", "breaker", "mcb", "socket", "inverter", "fan", "circuit", "appliance"],
    "HVAC": ["ac", "air condition", "cooling", "hvac", "compressor", "gas refill", "filter", "service", "ventilation"],
    "Carpentry": ["carpentry", "wood", "door", "hinge", "furniture", "cabinet", "shelf", "woodwork"],
    "Painting": ["paint", "wall", "coating", "finishing", "renovation", "interior", "decor"],
    "Locksmith": ["lock", "unlock", "key", "security", "bolt"],
    "Handyman": ["repair", "fix", "maintenance", "general"],
}


@dataclass(frozen=True)
class ScoredExpert:
    candidate: object
    expert_id: int
    full_name: str
    primary_expertise: str
    score: float
    breakdown: Dict[str, float]
    reasons: List[str]


class MatchingService:
    def __init__(
        self,
        repository: Repository,
        cache: CacheProvider | None = None,
        shard_router: SpatialShardRouter | None = None,
        shard_store: ShardExpertStore | None = None,
    ) -> None:
        self.repository = repository
        self.cache = cache
        self.shard_router = shard_router
        self.shard_store = shard_store

    def recommend(
        self,
        problem_text: str,
        top_n: int = 5,
        required_experience_years: int | None = None,
        *,
        latitude: float | None = None,
        longitude: float | None = None,
        radius_km: float = 15.0,
    ) -> Dict[str, object]:
        location_enabled = latitude is not None and longitude is not None
        region_buckets = None
        if location_enabled and self.shard_router is not None:
            region_buckets = self.shard_router.buckets_for_radius(latitude, longitude, radius_km)

        def within_radius(candidate) -> bool:
            if not location_enabled:
                return True
            if candidate.latitude is None or candidate.longitude is None:
                return False
            return haversine_km(latitude, longitude, candidate.latitude, candidate.longitude) <= radius_km

        if self.shard_store is not None:
            shard_result = self.shard_store.query_experts(
                available_only=True,
                search=None,
                primary_expertise=None,
                expertise_area=None,
                serves_as_resident=None,
                min_years_experience=None,
                max_years_experience=None,
                region_buckets=region_buckets,
                page=1,
                page_size=100000,
            )
            signature = shard_result["signature"]
        else:
            signature = self.repository.expert_catalog_signature(available_only=True, region_buckets=region_buckets)
        cache_key = self._cache_key(
            "expert-match",
            {
                "problem_text": problem_text,
                "top_n": top_n,
                "required_experience_years": required_experience_years,
                "latitude": latitude,
                "longitude": longitude,
                "radius_km": radius_km,
                "signature": signature,
            },
        )
        if self.cache is not None:
            cached = self.cache.get(cache_key)
            if isinstance(cached, dict):
                return cached

        if self.shard_store is not None:
            shard_rows = self.shard_store.query_experts(
                available_only=True,
                search=None,
                primary_expertise=None,
                expertise_area=None,
                serves_as_resident=None,
                min_years_experience=None,
                max_years_experience=None,
                region_buckets=region_buckets,
                page=1,
                page_size=100000,
            )["items"]
            candidates = [self._candidate_from_dict(item) for item in shard_rows]
        else:
            candidates = self.repository.expert_rows(available_only=True, region_buckets=region_buckets)
        candidates = [candidate for candidate in candidates if within_radius(candidate)]
        request_terms = self._infer_terms(problem_text)
        scored = [
            self._score_candidate(
                candidate,
                request_terms,
                required_experience_years,
                latitude=latitude,
                longitude=longitude,
                radius_km=radius_km,
            )
            for candidate in candidates
        ]
        ranked = sorted(scored, key=lambda item: item.score, reverse=True)[:top_n]
        response = {
            "request_text": problem_text,
            "suggestions": [
                {
                    "expert_id": item.candidate.expert_id,
                    "user_id": item.candidate.user_id,
                    "full_name": item.candidate.full_name,
                    "primary_expertise": item.candidate.primary_expertise,
                    "years_of_experience": item.candidate.years_of_experience,
                    "bio": getattr(item.candidate, "bio", None),
                    "available": item.candidate.available,
                    "serves_as_resident": item.candidate.serves_as_resident,
                    "expertise_areas": item.candidate.expertise_areas,
                    "score": round(item.score, 4),
                    "breakdown": {key: round(value, 4) for key, value in item.breakdown.items()},
                    "reasons": item.reasons,
                }
                for item in ranked
            ],
        }
        if self.cache is not None:
            self.cache.set(cache_key, response, ttl_seconds=settings.matching_cache_ttl_seconds)
        return response

    def _infer_terms(self, problem_text: str) -> List[str]:
        normalized = problem_text.lower()
        inferred: List[str] = []
        for category, keywords in KEYWORD_MAP.items():
            if any(keyword in normalized for keyword in keywords):
                inferred.append(category)

        tokens = set(re.findall(r"[a-zA-Z0-9-]+", normalized))
        inferred.extend(token.title() for token in tokens if len(token) > 4)
        return sorted(set(inferred))

    def _score_candidate(
        self,
        candidate,
        request_terms: List[str],
        required_experience_years: int | None,
        *,
        latitude: float | None,
        longitude: float | None,
        radius_km: float,
    ) -> ScoredExpert:
        expertise_tokens = {candidate.primary_expertise.lower(), *[skill.lower() for skill in candidate.expertise_areas]}
        term_tokens = {term.lower() for term in request_terms}

        overlap = len(expertise_tokens.intersection(term_tokens))
        request_scale = max(len(term_tokens), 1)
        skill_score = overlap / request_scale

        required = required_experience_years if required_experience_years is not None else 5
        experience_score = min(candidate.years_of_experience / max(required, 1), 1.0)

        availability_score = 1.0 if candidate.available else 0.0
        response_score = max(0.0, 1.0 - (candidate.avg_response_time_sec / 1200.0))

        reputation_base = (candidate.avg_rating / 5.0) if candidate.avg_rating else 0.0
        reputation_bonus = (candidate.acceptance_rate / 100.0) * 0.5 + (candidate.completion_rate / 100.0) * 0.5
        reputation_score = min(1.0, (reputation_base * 0.6) + (reputation_bonus * 0.4))
        location_score = 0.0
        if (
            latitude is not None
            and longitude is not None
            and candidate.latitude is not None
            and candidate.longitude is not None
        ):
            distance = haversine_km(latitude, longitude, candidate.latitude, candidate.longitude)
            location_score = max(0.0, 1.0 - (distance / max(radius_km, 1.0)))

        total = (
            0.45 * skill_score
            + 0.25 * experience_score
            + 0.15 * availability_score
            + 0.10 * reputation_score
            + 0.05 * response_score
        )
        total += 0.05 * location_score

        reasons = []
        matched_terms = sorted(expertise_tokens.intersection(term_tokens))
        if matched_terms:
            reasons.append(f"Skill overlap: {', '.join(matched_terms)}")
        reasons.append(f"{candidate.years_of_experience} years experience")
        reasons.append("Currently available" if candidate.available else "Temporarily unavailable")
        reasons.append(f"Rating {candidate.avg_rating:.1f}/5")
        if candidate.region_bucket:
            reasons.append(f"Assigned to shard bucket {candidate.region_bucket}")

        return ScoredExpert(
            candidate=candidate,
            expert_id=candidate.expert_id,
            full_name=candidate.full_name,
            primary_expertise=candidate.primary_expertise,
            score=total,
            breakdown={
                "skill": skill_score,
                "experience": experience_score,
                "availability": availability_score,
                "reputation": reputation_score,
                "response": response_score,
                "location": location_score,
            },
            reasons=reasons,
        )

    @staticmethod
    def _cache_key(prefix: str, payload: Dict[str, object]) -> str:
        encoded = json.dumps(payload, sort_keys=True, default=str).encode("utf-8")
        digest = hashlib.sha256(encoded).hexdigest()
        return f"{prefix}:{digest}"

    def _candidate_from_dict(self, item: Dict[str, object]):
        from types import SimpleNamespace

        return SimpleNamespace(
            expert_id=int(item["expert_id"]),
            user_id=int(item["user_id"]),
            full_name=str(item["full_name"]),
            email=str(item["email"]),
            primary_expertise=str(item["primary_expertise"]),
            years_of_experience=int(item["years_of_experience"]),
            bio=item.get("bio"),
            available=bool(item.get("available", False)),
            serves_as_resident=bool(item.get("serves_as_resident", False)),
            expertise_areas=list(item.get("expertise_areas", [])),
            avg_rating=float(item.get("avg_rating", 0.0) or 0.0),
            total_jobs=int(item.get("total_jobs", 0) or 0),
            acceptance_rate=float(item.get("acceptance_rate", 0.0) or 0.0),
            completion_rate=float(item.get("completion_rate", 0.0) or 0.0),
            cancellation_rate=float(item.get("cancellation_rate", 0.0) or 0.0),
            avg_response_time_sec=int(item.get("avg_response_time_sec", 0) or 0),
            latitude=item.get("latitude"),
            longitude=item.get("longitude"),
            city=item.get("city"),
            region_bucket=item.get("region_bucket"),
            shard_id=item.get("shard_id"),
        )

File: backend/src/main/java/com/example/backend/matching/service/MatchingService.java
package com.example.backend.matching.service;

import com.example.backend.auth.entity.ExpertProfile;
import com.example.backend.auth.repository.ExpertProfileRepository;
import com.example.backend.matching.dto.MatchingRequest;
import com.example.backend.matching.dto.MatchingResponse;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class MatchingService {

  private static final Pattern TOKEN_PATTERN = Pattern.compile("[a-zA-Z0-9-]+");

  private static final Map<String, List<String>> KEYWORD_MAP = Map.of(
      "Plumbing", List.of("plumb", "pipe", "leak", "drain", "faucet", "tap", "sink", "pipeline", "bathroom"),
      "Electrical", List.of("electrical", "wire", "wiring", "switch", "breaker", "mcb", "socket", "inverter", "fan", "circuit", "appliance"),
      "HVAC", List.of("ac", "air condition", "cooling", "hvac", "compressor", "gas refill", "filter", "service", "ventilation"),
      "Carpentry", List.of("carpentry", "wood", "door", "hinge", "furniture", "cabinet", "shelf", "woodwork"),
      "Painting", List.of("paint", "wall", "coating", "finishing", "renovation", "interior", "decor"),
      "Locksmith", List.of("lock", "unlock", "key", "security", "bolt"),
      "Handyman", List.of("repair", "fix", "maintenance", "general")
  );

  private final ExpertProfileRepository expertProfileRepository;

  public MatchingService(ExpertProfileRepository expertProfileRepository) {
    this.expertProfileRepository = expertProfileRepository;
  }

  @Transactional(readOnly = true)
  public MatchingResponse recommend(MatchingRequest request) {
    List<ExpertProfile> candidates = expertProfileRepository.findAllByAvailableTrue();

    boolean locationEnabled = request.latitude() != null && request.longitude() != null;
    double radiusKm = 15.0;

    List<ExpertProfile> filtered = candidates.stream()
        .filter(candidate -> !locationEnabled || withinRadius(candidate, request.latitude(), request.longitude(), radiusKm))
        .toList();

    Set<String> requestTerms = inferTerms(request.problemText());
    List<ExpertProfile> termFiltered = filterByTerms(filtered, requestTerms);

    List<MatchingResponse.Suggestion> scored = termFiltered.stream()
        .map(candidate -> scoreCandidate(candidate, requestTerms, request.latitude(), request.longitude(), radiusKm))
        .filter(s -> s.score() > 0.0)
        .sorted(Comparator.comparing(MatchingResponse.Suggestion::score).reversed())
        .toList();

    List<MatchingResponse.Suggestion> result = scored;
    if (request.topN() != null && request.topN() > 0) {
      result = scored.stream().limit(request.topN()).toList();
    }

    return new MatchingResponse(request.problemText(), result);
  }

  private boolean withinRadius(ExpertProfile candidate, Double lat, Double lon, double radiusKm) {
    if (candidate.getLatitude() == null || candidate.getLongitude() == null) {
      return false;
    }
    return haversineKm(lat, lon, candidate.getLatitude(), candidate.getLongitude()) <= radiusKm;
  }

  private Set<String> inferTerms(String problemText) {
    String normalized = problemText.toLowerCase();
    Set<String> inferred = new LinkedHashSet<>();

    for (Map.Entry<String, List<String>> entry : KEYWORD_MAP.entrySet()) {
      for (String keyword : entry.getValue()) {
        if (normalized.contains(keyword)) {
          inferred.add(entry.getKey());
          break;
        }
      }
    }

    var matcher = TOKEN_PATTERN.matcher(normalized);
    while (matcher.find()) {
      String token = matcher.group();
      if (token.length() > 4) {
        inferred.add(capitalize(token));
      }
    }

    return inferred;
  }

  private static String capitalize(String word) {
    if (word == null || word.isEmpty()) {
      return word;
    }
    return Character.toUpperCase(word.charAt(0)) + word.substring(1).toLowerCase();
  }

  private List<ExpertProfile> filterByTerms(List<ExpertProfile> candidates, Set<String> requestTerms) {
    if (requestTerms == null || requestTerms.isEmpty()) {
      return candidates;
    }

    Set<String> normalizedTerms = requestTerms.stream()
        .map(String::toLowerCase)
        .collect(Collectors.toSet());

    List<ExpertProfile> filtered = new java.util.ArrayList<>();
    for (ExpertProfile candidate : candidates) {
      Set<String> candidateTerms = new LinkedHashSet<>();
      candidateTerms.add(candidate.getPrimaryExpertise().toLowerCase());
      if (candidate.getExpertiseAreas() != null) {
        for (String skill : candidate.getExpertiseAreas()) {
          candidateTerms.add(skill.toLowerCase());
        }
      }

      String searchable = String.join(" ", candidateTerms);
      for (String term : normalizedTerms) {
        if (searchable.contains(term)) {
          filtered.add(candidate);
          break;
        }
      }
    }

    return filtered.isEmpty() ? candidates : filtered;
  }

  private MatchingResponse.Suggestion scoreCandidate(
      ExpertProfile candidate,
      Set<String> requestTerms,
      Double latitude,
      Double longitude,
      double radiusKm) {

    Set<String> expertiseTokens = new LinkedHashSet<>();
    expertiseTokens.add(candidate.getPrimaryExpertise().toLowerCase());
    if (candidate.getExpertiseAreas() != null) {
      for (String skill : candidate.getExpertiseAreas()) {
        expertiseTokens.add(skill.toLowerCase());
      }
    }

    Set<String> termTokens = requestTerms.stream()
        .map(String::toLowerCase)
        .collect(Collectors.toSet());

    double primaryScore = 0.0;
    double secondaryScore = 0.0;
    String primaryLower = candidate.getPrimaryExpertise().toLowerCase();

    for (String term : termTokens) {
      boolean primaryMatches = primaryLower.contains(term) || term.contains(primaryLower);
      if (primaryMatches) {
        primaryScore += 1.0;
      } else {
        for (String expertise : expertiseTokens) {
          if (expertise.contains(term) || term.contains(expertise)) {
            secondaryScore += 1.0;
            break;
          }
        }
      }
    }

    int requestScale = Math.max(termTokens.size(), 1);
    primaryScore = Math.min(primaryScore / requestScale, 1.0);
    secondaryScore = Math.min(secondaryScore / requestScale, 1.0);
    double skillScore = 0.6 * primaryScore + 0.4 * secondaryScore;

    double experienceScore = Math.min(candidate.getYearsOfExperience() / 5.0, 1.0);
    double availabilityScore = Boolean.TRUE.equals(candidate.getAvailable()) ? 1.0 : 0.0;

    double locationScore = 0.0;
    if (latitude != null && longitude != null && candidate.getLatitude() != null && candidate.getLongitude() != null) {
      double distance = haversineKm(latitude, longitude, candidate.getLatitude(), candidate.getLongitude());
      locationScore = Math.max(0.0, 1.0 - (distance / Math.max(radiusKm, 1.0)));
    }

    double total = 0.45 * skillScore + 0.25 * experienceScore + 0.15 * availabilityScore + 0.15 * locationScore;

    List<String> reasons = new java.util.ArrayList<>();
    if (primaryScore > 0.0) {
      reasons.add("Primary expertise matches query");
    } else if (secondaryScore > 0.0) {
      reasons.add("Secondary skills match query");
    }
    reasons.add(candidate.getYearsOfExperience() + " years experience");
    reasons.add(Boolean.TRUE.equals(candidate.getAvailable()) ? "Currently available" : "Temporarily unavailable");
    if (locationScore > 0.0) {
      reasons.add("Nearby service area");
    }

    Map<String, Double> breakdown = new LinkedHashMap<>();
    breakdown.put("skill", round(skillScore));
    breakdown.put("experience", round(experienceScore));
    breakdown.put("availability", round(availabilityScore));
    breakdown.put("location", round(locationScore));

    return new MatchingResponse.Suggestion(
        candidate.getId(),
        candidate.getUser().getId(),
        candidate.getUser().getFullName(),
        candidate.getPrimaryExpertise(),
        candidate.getYearsOfExperience(),
        candidate.getAvailable(),
        candidate.getServiceArea(),
        candidate.getLatitude(),
        candidate.getLongitude(),
        candidate.getExpertiseAreas(),
        round(total),
        breakdown,
        reasons);
  }

  private static double round(double value) {
    return Math.round(value * 10000.0) / 10000.0;
  }

  private static double haversineKm(double lat1, double lon1, double lat2, double lon2) {
    final double R = 6371.0;
    double dLat = Math.toRadians(lat2 - lat1);
    double dLon = Math.toRadians(lon2 - lon1);
    double a = Math.sin(dLat / 2) * Math.sin(dLat / 2)
        + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
        * Math.sin(dLon / 2) * Math.sin(dLon / 2);
    double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c;
  }
}
